Bugfix/562

MarkCuypersPpw opened the pull request:
fixes #562

- package-lock file was updated.
- Keyboard navigation fixes for arrow up/down skipping items.
- Keyboard enter-key fixes for multiselect dropdowns where the checkbox couldn't be (un)checked anymore.

Manual testing is advised.
Claude AI was used to help me understand what was happening and make suggestions on how to fix the code.

MarkCuypersPpw committed:
Update package-lock file

MarkCuypersPpw committed:
Fix issue with keyboard navigation using the arrow keys where options are skipped.

MarkCuypersPpw committed:
Fix issue with keyboard navigation when multiselect is on and the enter-key did not (un)check the checkboxes in the options items

macjohnny reviewed (commented):
thanks for the fix!

- did you test this extensively?
- is this backwards compatible with angular <=20? if thats not clear, we can also cut a new major version that requires angular >=21.

MarkCuypersPpw commented:
I copy/pasted the fixed component in 2 different projects:
- An Angular 21 project
- An Angular 19 project

In both projects I performed some manual testing for single select and multiple select. I tested keyboard navigation and it feels natural now. My console was opened while testing and no error appear. I don't know about edge cases, so it still might be useful if  you give it a try yourself. From what I see, the component works fine.

macjohnny commented:
thanks for testing!

did you dig into the code to see why there is the double stepping with angular 21? it seems a bit odd that we need to implement the keyboard navigation manually here, since we don't really deal with that in the component. if there is no simpler way to fix this odd behavior, i could be fine to just create a new major version requiring angular 21+, to avoid accidentally breaking users with angular <21.

MarkCuypersPpw commented:
I didn't dig into the code. I have no idea if there is a simpler way to fix this. There's clearly something changed with the events in Angular (Material?) causing this issue.

My suggestion is that this is an Angular 21 release. People using an older version of Angular (<= 20) can keep working with the stable v8.0.4.

MarkCuypersPpw commented:
Holidays are over and I finally found the time to dig into the code (with a colleague).

We came to the following conclusions:
- With Angular Material v21.0.0 it works fine, the issue was introduced in Angular Material v21.0.1
- This lead me to this PR: https://github.com/angular/components/pull/32363/files#diff-788cc1076f54d80b56b73ed0d3523a84f52133f4f6cdb423a919fa2ef3fc4b0b
- The addition of `cdkConnectedOverlayUsePopover="inline"` seems to break it.
- Inlining the popover activates the hostbinding of the mat-select also in the popover.
- This change causes the hostbinding on the keydown event in mat-select to fire https://github.com/angular/components/blob/main/src/material/select/select.ts (line 184) while it didn't fire before that change.
- Because that event fires, the _handleKeydown function is executed via the hostbinding. That same function is also called in the HTML template of mat-select https://github.com/angular/components/blob/main/src/material/select/select.html (line 60).
- With Angular Material v21.0.0 the _handleKeydown function is only called from within the template. NOT via the hostbinding.

From what I see, there might be a fix or workaround in the latest Angular Material code:
https://github.com/angular/components/commit/2c4f4bf043953d5c612aa5ba15448c052c9e58fa#diff-788cc1076f54d80b56b73ed0d3523a84f52133f4f6cdb423a919fa2ef3fc4b0b
This commit allows disabling popovers for mat-select. It might be a solution for my problem without needing a fix in ngx-mat-select-search.

MarkCuypersPpw commented:
Yesterday (2026-01-08) Angular Material released v21.0.6.

This version contains the before mentioned PR: https://github.com/angular/components/pull/32610

I now switched my app to use v21.0.6 of Angular Material and added this provider
`{ provide: OVERLAY_DEFAULT_CONFIG, useValue: { usePopover: false } },`
to my global app config. That makes it work fine again with the latest version.

macjohnny commented:
@MarkCuypersPpw thanks for the detailed analysis and for following up! glad this is resolved now (or at least there is a config to fix this)
